### PR TITLE
Include `nn.Parameter` in ModelSummary

### LINKED ISSUE #12736: ModelSummary ignores nn.Parameter
## 🐛 Bug

I don't know if this is a bug of a feature, but currently ModelSummary only reports member `nn.Module`s and ignores `nn.Parameter`s that are directly created. I believe these parameters are an important part (and sometimes, a big part in terms of the number of weights) of the model so that they should be reported.

cc @borda @kaushikb11 @awaelchli @rohitgr7

### PR DESCRIPTION
**Pull request by yeonju7kim**
## What does this PR do?

When we defined torch.nn.parameter object in the model and then print ModelSummary for the model, the printed result doesn't show the parameter number of the torch.nn.parameter object.
Let me show you the example.

``` python
import torch

from pytorch_lightning import LightningModule
from pytorch_lightning.utilities.model_summary import ModelSummary

class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.param = torch.nn.Parameter(torch.rand(2, 2))  # 2 * 2 = 4
        self.layer = torch.nn.Linear(2, 2)  # 2 * 2 + 2 = 6


if __name__ == "__main__":
    model = BoringModel()
    summary = ModelSummary(model)
    print(summary)
```

```
  | Name  | Type   | Params
---------------------------------
0 | param | Parameter | 4                        <- I add this line. 
1 | layer | Linear    | 6          
---------------------------------
10        Trainable params
0         Non-trainable params
10        Total params
0.000     Total estimated model params size (MB)
```

Fixes #12736

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Yes! It's my first contribution. I welcome your feedback!

### CODE REVIEW

**Comment by yeonju7kim** [comment]
I should fix mypy error first.

File: src/pytorch_lightning/utilities/deepspeed_model_summary.py [hunk 1]
```diff
@@ -35,12 +36,24 @@ class DeepSpeedLayerSummary(LayerSummary):
     @property
     def num_parameters(self) -> int:
         """Returns the number of parameters in this module."""
-        return sum(deepspeed_param_size(p) if not _is_lazy_weight_tensor(p) else 0 for p in self._module.parameters())
+        if isinstance(self._module, torch.nn.Module):
+            return sum(
+                deepspeed_param_size(p) if not _is_lazy_weight_tensor(p) else 0 for p in self._module.parameters()
+            )
+        elif isinstance(self._module, torch.nn.Parameter):
+            return int(np.prod(self._module.shape))
```
Comment (justusschock) [inline src/pytorch_lightning/utilities/deepspeed_model_summary.py:44]: probably need deepspeed handling for this kind of parameter as well?

File: src/pytorch_lightning/utilities/deepspeed_model_summary.py [hunk 2]
```diff
@@ -35,12 +36,24 @@ class DeepSpeedLayerSummary(LayerSummary):
...
+            raise NotImplementedError
 
     @property
     def average_shard_parameters(self) -> int:
         """Returns the number of parameters in this module."""
-        return sum(p.partitioned_size() if not _is_lazy_weight_tensor(p) else 0 for p in self._module.parameters())
+        if isinstance(self._module, torch.nn.Module):
+            return sum(p.partitioned_size() if not _is_lazy_weight_tensor(p) else 0 for p in self._module.parameters())
+        elif isinstance(self._module, torch.nn.Parameter):
+            return int(np.prod(self._module.shape))
```
Comment (justusschock) [inline src/pytorch_lightning/utilities/deepspeed_model_summary.py:54]: this doesn't look right. should also return either the partitioned_size or 0

File: src/pytorch_lightning/utilities/model_summary.py [hunk 3]
```diff
@@ -241,7 +268,12 @@ def model_size(self) -> float:
         return self.total_parameters * self._precision_megabytes
 
     def summarize(self) -> Dict[str, LayerSummary]:
-        summary = OrderedDict((name, LayerSummary(module)) for name, module in self.named_modules)
+        all_parameter: List[Tuple[str, Union[nn.Module, nn.Parameter]]]
+        all_parameter = []
+        all_parameter.extend(self.named_modules)
+        all_parameter.extend(self.named_parameters)
```
Comment (justusschock) [inline src/pytorch_lightning/utilities/model_summary.py:274]: are you sure this isn't traversing recursively?

File: src/pytorch_lightning/utilities/deepspeed_model_summary.py [hunk 4]
```diff
@@ -35,12 +36,24 @@ class DeepSpeedLayerSummary(LayerSummary):
     @property
     def num_parameters(self) -> int:
         """Returns the number of parameters in this module."""
-        return sum(deepspeed_param_size(p) if not _is_lazy_weight_tensor(p) else 0 for p in self._module.parameters())
+        if isinstance(self._module, torch.nn.Module):
+            return sum(
+                deepspeed_param_size(p) if not _is_lazy_weight_tensor(p) else 0 for p in self._module.parameters()
+            )
+        elif isinstance(self._module, torch.nn.Parameter):
```
Comment (justusschock) [inline src/pytorch_lightning/utilities/deepspeed_model_summary.py:43]: I would handle this separately. The docstring says it handles the numver of parameters in **this module** so the module itself shouldn't be a nn.Parameter .

Also the class is named `LayerSummary`, to me that looks like directly registered parameters should be handled externally as well.

File: src/pytorch_lightning/utilities/model_summary.py [hunk 5]
```diff
@@ -120,9 +126,13 @@ def layer_type(self) -> str:
     @property
     def num_parameters(self) -> int:
         """Returns the number of parameters in this module."""
-        return sum(
-            cast(int, np.prod(p.shape)) if not _is_lazy_weight_tensor(p) else 0 for p in self._module.parameters()
-        )
+        if isinstance(self._module, nn.Module):
+            return sum(int(np.prod(p.shape)) if not _is_lazy_weight_tensor(p) else 0 for p in self._module.parameters())
+        elif isinstance(self._module, nn.Parameter):
```
Comment (justusschock) [inline src/pytorch_lightning/utilities/model_summary.py:131]: Same thing. IMO this doesn't belong in this class.

**Review by justusschock — COMMENTED** [review]
In general: What about buffers they may contribute to the memory footprint as well.

File: src/pytorch_lightning/utilities/model_summary.py (hunk 3 above)
Comment (yeonju7kim) [reply src/pytorch_lightning/utilities/model_summary.py:274]: I used 'reverse', so it will not traverse recursively.

File: src/pytorch_lightning/utilities/deepspeed_model_summary.py (hunk 4 above)
Comment (yeonju7kim) [reply src/pytorch_lightning/utilities/deepspeed_model_summary.py:43]: The class LayerSummary was originally the class that summarizes a single layer whose type was nn.Module.
However, not to ignore nn.Parameter I changed the type of layer to be nn.Module or nn.Parameter.
Because it can be nn.Module and nn.Parameter, I should write as above.

I think I should make another class that summarizes nn.Parameter and it will be class ParameterSummary.
